### PR TITLE
[SPARK-47348][INFRA] Align the version of the ubuntu host machine where the GitHub workflow runs on

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
- The pr aims to align `the version` of the ubuntu host machine where the GitHub workflow runs on.
- In all the `YML` files in the workflow, only `two` place `retains` the Ubuntu version of `20.04`, because `databricks/tpcds-kit` depends on it.
https://github.com/apache/spark/blob/6c30525e0d71c707aaaed1be49600b866bced49b/.github/workflows/build_and_test.yml#L857-L858
https://github.com/apache/spark/blob/6c30525e0d71c707aaaed1be49600b866bced49b/.github/workflows/benchmark.yml#L63

### Why are the changes needed?
I found that some workflow run on `ubuntu-latest` (`ubuntu-22.04`), while others run on `ubuntu-20.04`.
Make the testing environment more consistent.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @HyukjinKwon

**Review by zhengruifeng — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
WDYT @Yikun and @dongjoon-hyun ?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side. Thank you, @HyukjinKwon and @panbingkun .

**Comment by panbingkun** [comment]
> BTW, we are assuming that `Ubuntu 24.04` next month in this PR, right?

Yeah, currently `ubuntu-latest` is `ubuntu-22.04`. With the latest stable images (eg.: `ubuntu-24.04`) that GitHub provides, we will use it (`ubuntu-24.04`).

https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources
image

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Hi, @panbingkun and all.

Unfortunately, this seems to break `branch-3.4` and `branch-3.5` CI due to the Python version and package differences. Do you think we can isolate this PR to `master` branch only?

- Python Linter Failure
    - https://github.com/apache/spark/actions/runs/8305693061/job/22733444259
```
starting mypy annotations test...
annotations failed mypy checks:
python/pyspark/pandas/plot/matplotlib.py:22: error: Module "matplotlib.axes._base" has no attribute "_process_plot_format"  [attr-defined]
Found 1 error in 1 file (checked 425 source files)
1
Error: Process completed with exit code 1.
```

- Pandas Failure
   - https://github.com/apache/spark/actions/runs/8305693061/job/22733444673
```
======================================================================
ERROR [15.158s]: test_aggregate (pyspark.pandas.tests.test_groupby_slow.GroupBySlowTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/pandas/tests/test_groupby_slow.py", line 287, in test_aggregate
    stats_pdf.sort_values(by=[(20, "min"), (20, "max"), (30, "sum")]).reset_index(
  File "/usr/local/lib/python3.9/dist-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/pandas/core/frame.py", line 6894, in sort_values
    keys = [self._get_label_or_level_values(x, axis=axis) for x in by]
  File "/usr/local/lib/python3.9/dist-packages/pandas/core/frame.py", line 6894, in <listcomp>
    keys = [self._get_label_or_level_values(x, axis=axis) for x in by]
  File "/usr/local/lib/python3.9/dist-packages/pandas/core/generic.py", line 1838, in _get_label_or_level_values
    self._check_label_or_level_ambiguity(key, axis=axis)
  File "/usr/local/lib/python3.9/dist-packages/pandas/core/generic.py", line 1780, in _check_label_or_level_ambiguity
    and key in self.axes[axis].names
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

----------------------------------------------------------------------
Ran 15 tests in 174.031s

FAILED (errors=1)
```

**Comment by dongjoon-hyun** [comment]
For the record, after testing #45545, I found that we need the following to recover `branch-3.5` which were the previous successful run.

```
matplotlib==3.7.2
torch==2.0.1
torchvision==0.15.2
```

**Comment by panbingkun** [comment]
Let me investigate it now.

**Comment by panbingkun** [comment]
For the record:

1.branch 3.4, Python Linter Failure, https://github.com/apache/spark/actions/runs/8305693061/job/22733444259, 
run on: `ubuntu  20.04.6`
image

**Comment by panbingkun** [comment]
From the case given above, the problems that `branch 3.5 run on ubuntu-22.04` & `branch 3.4 run on ubuntu-20.04 `encountered in workflow `Linters, licenses, dependencies and documentation generation` are the same, as shown below:
```

```

**Comment by dongjoon-hyun** [comment]
Yes. It looks like that. I have no clue about how this PR can affect the old branch CIs.
So, I'm investigating without reverting this PR.

**Comment by HyukjinKwon** [comment]
For linter failure at the scheduled job, I think it's because of different Python dependency versions. For example, at least I see one diff

```
Requirement already satisfied: typing-extensions>=3.7.4 in /usr/local/lib/python3.9/dist-packages (from mypy==0.920) (4.4.0)
```

https://github.com/apache/spark/actions/runs/8278032972/job/22649628764 (failed)

vs

```
Requirement already satisfied: typing-extensions>=3.7.4 in /usr/local/lib/python3.9/dist-packages (from mypy==0.920) (4.10.0)
```

https://github.com/apache/spark/actions/runs/8293729419/job/22699200117 (passed)

**Comment by HyukjinKwon** [comment]
The test failures in `branch-3.5` and `branch-3.4` seems because of the behaviour difference of pyarrow version.

```
pyarrow                  15.0.1
```

https://github.com/apache/spark/actions/runs/8305690576/job/22746538190 (failed)

vs

```
pyarrow                  12.0.1
```

https://github.com/apache/spark/actions/runs/8305690576/job/22746538190 (passed)

**Comment by HyukjinKwon** [comment]
For https://github.com/apache/spark/pull/45416#issuecomment-2002352417, it is because we use the cached image from the master branch. So we explicitly overwrite the Python dependencies for linter jobs (https://github.com/apache/spark/blob/master/.github/workflows/build_and_test.yml#L682-L688). TL;DR, for scheduled jobs, it uses the cached image for `master` branch.

For https://github.com/apache/spark/pull/45416#issuecomment-2002353686, I commented at https://github.com/apache/spark/pull/45545#issuecomment-2002355919. TL;DR, In this case, it uses the cached image for the branch itself.

**Comment by dongjoon-hyun** [comment]
Thank you for the analysis, @HyukjinKwon ! I closed the investigation PR, #45545 .

**Comment by panbingkun** [comment]
> For [#45416 (comment)](https://github.com/apache/spark/pull/45416#issuecomment-2002352417), it is because we use the cached image from the master branch. So we explicitly overwrite the Python dependencies for linter jobs (https://github.com/apache/spark/blob/master/.github/workflows/build_and_test.yml#L682-L688). TL;DR, for scheduled jobs, it uses the cached image for `master` branch. In this case, we should pin the dependencies at `master` branch to remove the version diff, cc @panbingkun can you followup this please?
> 
> For [#45416 (comment)](https://github.com/apache/spark/pull/45416#issuecomment-2002353686), I commented at [#45545 (comment)](https://github.com/apache/spark/pull/45545#issuecomment-2002355919). TL;DR, In this case, it uses the cached image for the branch itself. It should be fixed if we backport corresponding commits I pointed out in the comment.

Thank you very much for your explanation @HyukjinKwon
Okay, let me try.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -52,7 +52,7 @@ on:
 jobs:
   precondition:
     name: Check changes
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-latest
```
Comment (dongjoon-hyun) [inline .github/workflows/build_and_test.yml:55]: Here is my theory.

According to the above @HyukjinKwon analysis, this line could be the reason why this PR regenerates the cached images of branch-3.5 and branch-3.4.

`precondition` job is literally changed here although the actual version is the same, Ubuntu 22.04 == Ubuntu Latest. And, it  invokes the following.

https://github.com/apache/spark/blob/714feb7b5aec948c3499ad486dc63ea89a241ffa/.github/workflows/build_and_test.yml#L111-L118

If we look at `Starting job container` stage of `branch-3.5` job, `ghcr.io/apache/apache-spark-ci-image:branch-3.5-8276441035` image has different sha values. It seems to be regenerated.

**BEFORE**: https://github.com/apache/spark/actions/runs/8276441035/job/22645041254

```
Digest: sha256:c97bcef01a653038dba464bc5d947340be4385efa18808b4a3eb813095ea95f9
Status: Downloaded newer image for ghcr.io/apache/apache-spark-ci-image:branch-3.5-8276441035
```

**AFTER**: https://github.com/apache/spark/actions/runs/8276441035/job/22733550856
```
Digest: sha256:a91feba3ada152b513192acb4fa3b2930655e37644f6bad11c30d6c7a6a12383
Status: Downloaded newer image for ghcr.io/apache/apache-spark-ci-image:branch-3.5-8276441035
```

WDTY, @HyukjinKwon and @panbingkun ?

**Comment by panbingkun** [comment]
@HyukjinKwon 
In addition, I would like to ask for advice.  In order to solve the problem of `scheduled jobs`, after pinning the Python dependencies on the `master`, how can I easily test and verify them?  
So far, I seem to have not found a suitable method.

**Comment by HyukjinKwon** [comment]
What you're doing at https://github.com/apache/spark/pull/45551 would be the way I would test it.

**Comment by HyukjinKwon** [comment]
Explicitly provide those info specified in https://github.com/apache/spark/blob/master/.github/workflows/build_branch35.yml, and run it within your PR.

**Comment by panbingkun** [comment]
Okay, I'll continue with this PR https://github.com/apache/spark/pull/45551

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:55]: Yes, correctly. This only reveals the technical depth, @panbingkun .

**Pull request closed without merging**